feat(nextjs): can set compiler when setting up `nxComponentTestingPreset`

erenken opened the pull request:
Currently *nextjs* only uses the *swa* compiler in Cypress component testing.

To use @cypress/code-coverage you need to use *babel*, you can now select which compiler to use.

closed #19131

<!-- Please make sure you have read the submission guidelines before posting an PR -->
<!-- https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md#-submitting-a-pr -->

<!-- Please make sure that your commit message follows our format -->
<!-- Example: `fix(nx): must begin with lowercase` -->

## Current Behavior
Cypress Code Coverage doesn't work in a *next* project when using `nxComponentTestingPreset` from `@nx/next/plugins/component-testing`. Currently it works if you are using a *react* project and importing `nxComponentTestingPreset` from `@nx/react/plugins/component-testing`, but tests using NavLink fail because they are missing process with using this preset.

## Expected Behavior
I would expect Cypress Code Coverage to work in a *next* and *react* project.

## Related Issue(s)
<!-- Please link the issue being fixed so it gets closed when this is merged. -->

Fixes #19131 

Unable to run test on `devcontainer`, so that was updated to use latest version.

mandarini commented:
Hey @erenken thanks for this PR! Can you please add some more tests around this?

mandarini commented:
@erenken changes look good, we just need a few tests around this and the pipelines to pass! :D Just ping me when you're done and I will merge! Thank you for this PR!

erenken commented:
> @erenken changes look good, we just need a few tests around this and the pipelines to pass! :D Just ping me when you're done and I will merge! Thank you for this PR!

Are there any existing tests for the `nxComponentTestingPreset` method?  I was trying to find them, and I couldn't find anything that executes the code.  The only tests that have `nxComponentTestingPreset` in them are all doing contains or snapshot comparisons.  

Just adding the line:

`const result = nxComponentTestingPreset('myFile.json', {});`

Produces this error:

```typescript
    /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by /workspaces/nx/packages/nx/src/native/nx.linux-x64-gnu.node)

      180 |           try {
      181 |             if (localFileExisted) {
    > 182 |               nativeBinding = require('./nx.linux-x64-gnu.node')
          |                               ^
      183 |             } else {
      184 |               nativeBinding = require('@nx/nx-linux-x64-gnu')
      185 |             }

      at CompileFunctionRuntime._loadModule (../../node_modules/.pnpm/jest-runtime@29.5.0/node_modules/jest-runtime/build/index.js:1009:29)
      at Object.<anonymous> (../nx/src/native/index.js:182:31)
      at Object.<anonymous> (../nx/src/hasher/task-hasher.ts:23:1)
      at Object.<anonymous> (../nx/src/hasher/hash-task.ts:4:1)
      at Object.<anonymous> (../nx/src/tasks-runner/task-orchestrator.ts:21:1)
      at Object.<anonymous> (../nx/src/tasks-runner/default-tasks-runner.ts:2:1)
      at Object.<anonymous> (../devkit/nx-reexports-pre16.ts:209:1)
      at Object.<anonymous> (../devkit/index.js:21:20)
      at Object.<anonymous> (../cypress/src/utils/ct-helpers.ts:2:1)
      at Object.<anonymous> (plugins/component-testing.ts:1:1)
      at Object.<anonymous> (plugins/component-testing.spec.ts:1:1)
```

When I comment out that line no errors and the tests pass.  Any recommendations?

erenken commented:
Never mind, I got it all working after updating the devcontainer.  Pushing tests soon.

mandarini commented:
Thank you very much for this PR @erenken !

barbados-clemens reviewed (changes requested):
Thanks for your PR. There isn't a need to add those deps to the package.json since the tests for component testing should be in the e2e projects where they will be installed at run time in the testing project vs the actual nx repo. 

thanks for updating the dev container.

erenken commented on packages/next/project.json:
```diff
@@ -59,6 +59,12 @@
+      "options": {
+        "cypressConfig": "cypress-component-test.config.ts"
+      }
+    }
```
This has been removed.

erenken commented on package.json:
```diff
@@ -149,6 +149,7 @@
     "conventional-changelog-cli": "^2.0.23",
     "convert-source-map": "^2.0.0",
     "copy-webpack-plugin": "^10.2.4",
+    "css-loader": "^6.8.1",
```
This has been removed.

erenken commented on package.json:
```diff
@@ -237,6 +238,7 @@
     "parse5": "4.0.0",
     "postcss": "8.4.19",
     "postcss-import": "~14.1.0",
+    "postcss-loader": "^7.3.3",
```
This has been removed.

erenken commented on packages/next/plugins/component-testing.spec.ts:
```diff
@@ -0,0 +1,62 @@
+import { NxComponentTestingOptions } from '@nx/cypress/plugins/cypress-preset';
+import { nxComponentTestingPreset } from './component-testing';
+
+describe('nxComponentTestingPreset', () => {
```
e2e tests have been added, and are all passing, well at least locally :)

erenken commented on packages/react/plugins/component-testing/index.ts:
```diff
@@ -160,7 +160,7 @@ export function nxComponentTestingPreset(
     webpackConfig = buildBaseWebpackConfig({
       tsConfigPath: findTsConfig(normalizedProjectRootPath),
-      compiler: 'babel',
+      compiler: options?.compiler || 'babel',
```
Done

erenken commented:
FYI ci/circleci: main-linux is giving this error:

`Error when connecting to Nx Cloud. Code: 500. Error: unknown error`

There appears to be a problem with the environment.

erenken commented:
@mandarini I got all the changes that you and @barbados-clemens wanted.  I removed the packages I added to packages.json and created E2E test, and everything is passing.  If I missed something or you need more, please let me know.  I would like to get this merged in so we can get code coverage working in our project.

barbados-clemens commented on .devcontainer/devcontainer.json:
```diff
@@ -3,7 +3,7 @@
 {
   "name": "NxDevContainer",
   // Or use a Dockerfile or Docker Compose file. More info: https://containers.dev/guide/dockerfile
-  "image": "mcr.microsoft.com/devcontainers/typescript-node:0-18",
```
nitpick, let's not have this change in this PR, also probs shouldn't be latest as could end up using version that isn't fully supported yet. 

Mainly just doing want these changes to end up in PRs that aren't fully related. Sorry didn't catch this initially.

erenken replied:
Not changing it now, but my thought would be to change it to `0-20`

Changes made.

barbados-clemens commented on pnpm-lock.yaml:
```diff
@@ -533,6 +533,9 @@ devDependencies:
   copy-webpack-plugin:
     specifier: ^10.2.4
     version: 10.2.4(webpack@5.88.0)
+  css-loader:
```
should revert these changes from lock file and package.json.

erenken replied:
Sorry, I thought I got all of them.  I just revered the files to what was in master.

erenken replied:
OK, sorry for all of the commits, was having issues with my branches being out of sync.  package and pnpm-lock are both revered.  We should be good now.

barbados-clemens reviewed (commented):
looks like the package.json changes are still there can you revert those real quick?

erenken commented:
Look like the `ci/circleci: main-linux` needs to be re-ran it is giving the error:

```
Failed to install browsers
Error: Failed to install chrome
```

erenken commented:
I know it looks like the main-linux failed, but I think it just needs to be re-ran.  I have found that these pipelines are very temperamental and the only way to re-run them is to commit a change or update from master.  It was working before the last update from master.
